rust/datafusion/src/datasource/parquet.rs

```diff
@@ -0,0 +1,570 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+//! Parquet Data source
+
+use std::fs::File;
+use std::string::String;
+use std::sync::{Arc, Mutex};
+
+use arrow::array::{Array, PrimitiveArray};
+use arrow::builder::{BinaryBuilder, PrimitiveBuilder, TimestampNanosecondBuilder};
+use arrow::datatypes::*;
+use arrow::record_batch::RecordBatch;
+
+use parquet::column::reader::*;
+use parquet::data_type::{ByteArray, Int96};
+use parquet::file::reader::*;
+use parquet::reader::schema::parquet_to_arrow_schema;
+
+use crate::datasource::{RecordBatchIterator, ScanResult, Table};
+use crate::execution::error::{ExecutionError, Result};
+
+pub struct ParquetTable {
+    filename: String,
+    schema: Arc<Schema>,
+}
+
+impl ParquetTable {
+    pub fn try_new(filename: &str) -> Result<Self> {
```

Comment:
nit: Maybe `filename` should be of type `AsRef<Path>` just like `File::open`?  (probably does not need to be included in this PR)

Reply:
Sure. I'd prefer to look at that as a separate PR as you suggested.